elements/research/ddiotune.cc

```diff
@@ -0,0 +1,385 @@
+	printf("Use_Allocating_Flow_Wr val: 0x%" PRIx32 "\n",val&use_allocating_flow_wr_MASK);
+	if(val&use_allocating_flow_wr_MASK)
+		return 1;
+	else
+		return 0;
+}
+
+/* Enable DDIO for the input NIC */
+void
+DDIOTune::ddio_enable(uint8_t nic_bus)
+{
+	uint32_t val;
+	if(!pacc)
+		init_pci_access();
+
+	if(!ddio_status(nic_bus))
+	{
+		struct pci_dev* dev=find_ddio_device(nic_bus);
+		if(!dev){
+			printf("No device found!\n");
+			exit(1);
+		}
+		val=pci_read_long(dev,PERFCTRLSTS_0);
+		pci_write_long(dev,PERFCTRLSTS_0,val|use_allocating_flow_wr_MASK);
+		printf("DDIO is enabled!\n");
+	} else
+	{
+		printf("DDIO was already enabled!\n");
+	}
+}
+
+/* Disable DDIO for the input NIC */
+void
+DDIOTune::ddio_disable(uint8_t nic_bus)
+{
+	uint32_t val;
+	if(!pacc)
+		init_pci_access();
+
+	if(ddio_status(nic_bus))
+	{
+		struct pci_dev* dev=find_ddio_device(nic_bus);
+		if(!dev){
+			printf("No device found!\n");
+			exit(1);
+		}
+		val=pci_read_long(dev,PERFCTRLSTS_0);
+		pci_write_long(dev,PERFCTRLSTS_0,val&(~use_allocating_flow_wr_MASK));
+		printf("DDIO is disabled!\n");
+	} else 
+	{
+		printf("DDIO was already disabled\n");
+	}
+}
+
+/* Print PCI root information */
+void
+DDIOTune::print_dev_info(struct pci_dev *dev)
+{
+	if(!dev){
+		printf("No device found!\n");
+		exit(1);
+	}
+	unsigned int c;
+	char namebuf[1024], *name;
+	printf("========================\n");
+	printf("%04x:%02x:%02x.%d vendor=%04x device=%04x class=%04x irq=%d (pin %d) base0=%lx \n",
+		dev->domain, dev->bus, dev->dev, dev->func, dev->vendor_id, dev->device_id,
+		dev->device_class, dev->irq, c, (long) dev->base_addr[0]);
+	name = pci_lookup_name(pacc, namebuf, sizeof(namebuf), PCI_LOOKUP_DEVICE, dev->vendor_id, dev->device_id);
+	printf(" (%s)\n", name);
+	printf("========================\n");
+}
+
+/* Constructor */
+DDIOTune::DDIOTune() : _w(-1)
+{
+	eax=0;ebx=0;ecx=0;edx=0;
+	_numa_node=-1;
+	_msr_cpu=-1;
+	IIO_LLC_WAYS_VALUE_NEW=0;
+	ways = new uint32_t[CACHE_LEVELS];
+	partitions = new uint32_t[CACHE_LEVELS];
+	linesize = new uint32_t[CACHE_LEVELS];
+	sets = new uint32_t[CACHE_LEVELS];
+	size = new uint32_t[CACHE_LEVELS];
+
+	init_pci_access();
+}
+
+/* Destructor */
+DDIOTune::~DDIOTune()
+{
+	delete ways, partitions, linesize, sets, size;
+
+	if(_w) {
+		/*Revert IIO_LLC_WAYS value*/
+		int32_t ret = wrmsr(IIO_LLC_WAYS_REGISTER, &IIO_LLC_WAYS_VALUE,_msr_cpu);
+		if (ret <0)
+			click_chatter("Could not revert IIO_LLC_WAYS value!");
+		click_chatter("IIO_LLC_WAYS value changed back to 0x%x from 0x%x",IIO_LLC_WAYS_VALUE, IIO_LLC_WAYS_VALUE_NEW);
+	}
+	/* Invert DDIO state*/
+	if(_ddio_state) {
+		ddio_enable(_pci_dev);
+		click_chatter("DDIO state is reverted back to enabled!");
+	} else {
+		ddio_disable(_pci_dev);
+		click_chatter("DDIO state is reverted back to disabled!");
+	}
+	pci_cleanup(pacc);		/* Close everything */
+}
+
+
+/* Configure: main function */
+int
+DDIOTune::configure(Vector<String> &conf, ErrorHandler *errh)
+{
```

Comment:
Why this removal? But if you uncomment don't forget the instance(0) in the constructor.

Reply:
I fixed it based on Georgios's comment.